File: backend/src/services/inventory/serpapi/serpApiProvider.ts
import axios from 'axios';
import type {
  FlightProvider,
  HotelProvider,
  FlightQuery,
  FlightOffer,
  HotelQuery,
  HotelOffer,
  AirportRef,
} from '../types';

const SERPAPI_BASE = 'https://serpapi.com';

/**
 * SerpApi provider — implements FlightProvider + HotelProvider
 * using Google Flights and Google Hotels engines.
 */
export class SerpApiProvider implements FlightProvider, HotelProvider {
  private apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  // ─── FlightProvider ───

  async searchOffers(query: FlightQuery): Promise<FlightOffer[]> {
    const params: Record<string, string> = {
      engine: 'google_flights',
      api_key: this.apiKey,
      departure_id: query.origin,
      arrival_id: query.destination,
      outbound_date: query.departureDate,
      type: query.returnDate ? '2' : '1',
      adults: String(query.adults),
      children: String(query.children),
      travel_class: query.travelClass,
      currency: query.currency,
    };

    if (query.returnDate) {
      params.return_date = query.returnDate;
    }

    if (query.deepSearch) {
      params.deep_search = 'true';
    }

    if (query.maxPrice) {
      params.max_price = String(query.maxPrice);
    }

    const { data } = await axios.get(`${SERPAPI_BASE}/search`, { params });

    if (data.search_metadata?.status === 'Error') {
      throw new Error(`SerpApi error: ${data.error || 'Unknown error'}`);
    }

    const offers: FlightOffer[] = [];

    for (const flight of data.best_flights || []) {
      offers.push(this.normalizeFlightOffer(flight, true, query.currency));
    }

    for (const flight of data.other_flights || []) {
      offers.push(this.normalizeFlightOffer(flight, false, query.currency));
    }

    return offers;
  }

  async autocomplete(term: string): Promise<AirportRef[]> {
    const { data } = await axios.get(`${SERPAPI_BASE}/search`, {
      params: {
        engine: 'google_flights_autocomplete',
        api_key: this.apiKey,
        q: term,
      },
    });

    if (data.search_metadata?.status === 'Error') {
      throw new Error(`SerpApi error: ${data.error || 'Unknown error'}`);
    }

    return (data.suggestions || []).map((s: any) => ({
      code: s.iata_code,
      name: s.name,
      city: s.type === 'City' ? s.name : undefined,
    }));
  }

  // ─── HotelProvider ───

  async search(query: HotelQuery): Promise<HotelOffer[]> {
    const sortMap: Record<string, string> = {
      'relevance': '0',
      'price-low': '3',
      'rating-high': '8',
      'reviews': '13',
    };

    const params: Record<string, string> = {
      engine: 'google_hotels',
      api_key: this.apiKey,
      q: query.destination,
      check_in_date: query.checkIn,
      check_out_date: query.checkOut,
      adults: String(query.adults),
      children: String(query.children),
      currency: query.currency,
      sort: sortMap[query.sort] || '0',
    };

    if (query.minPrice) {
      params.min_price = String(query.minPrice);
    }

    if (query.maxPrice) {
      params.max_price = String(query.maxPrice);
    }

    const { data } = await axios.get(`${SERPAPI_BASE}/search`, { params });

    if (data.search_metadata?.status === 'Error') {
      throw new Error(`SerpApi error: ${data.error || 'Unknown error'}`);
    }

    return (data.properties || []).map((p: any) => this.normalizeHotelOffer(p, query.currency));
  }

  // ─── Private normalizers ───

  private normalizeFlightOffer(raw: any, isBest: boolean, currency: string): FlightOffer {
    const legs = (raw.flights || []).map((f: any) => ({
      departureAirport: {
        code: f.departure_airport?.id || '',
        name: f.departure_airport?.name || '',
        time: f.departure_airport?.time,
      },
      arrivalAirport: {
        code: f.arrival_airport?.id || '',
        name: f.arrival_airport?.name || '',
        time: f.arrival_airport?.time,
      },
      airline: f.airline || '',
      airlineLogo: f.airline_logo,
      flightNumber: f.flight_number || '',
      airplane: f.airplane,
      travelClass: f.travel_class,
      duration: f.duration || 0,
      legroom: f.legroom,
      overnight: f.overnight,
      extensions: f.extensions,
    }));

    const layovers = (raw.layovers || []).map((l: any) => ({
      airportCode: l.id || '',
      airportName: l.name,
      duration: l.duration || 0,
      overnight: l.overnight,
    }));

    const id = `flight_${raw.booking_token || Math.random().toString(36).slice(2)}`;

    return {
      id,
      legs,
      layovers,
      totalDuration: raw.total_duration || 0,
      price: raw.price || 0,
      currency,
      type: raw.type === 'round-trip' ? 'round-trip' : 'one-way',
      bookingToken: raw.booking_token,
      carbonEmissions: raw.carbon_emissions
        ? {
            thisFlight: raw.carbon_emissions.this_flight,
            typicalForRoute: raw.carbon_emissions.typical_for_this_route,
            differencePercent: raw.carbon_emissions.difference_percent,
          }
        : undefined,
      isBest,
    };
  }

  private normalizeHotelOffer(raw: any, currency: string): HotelOffer {
    const id = `hotel_${raw.property_token || Math.random().toString(36).slice(2)}`;

    return {
      id,
      name: raw.name || '',
      type: raw.type,
      ratePerNight: raw.rate_per_night?.extracted_lowest,
      totalRate: raw.total_rate?.extracted_lowest,
      currency,
      ratePerNightFormatted: raw.rate_per_night?.lowest,
      totalRateFormatted: raw.total_rate?.lowest,
      rating: raw.rating,
      reviewsCount: raw.reviews,
      gpsCoordinates: raw.gps_coordinates
        ? {
            latitude: raw.gps_coordinates.latitude,
            longitude: raw.gps_coordinates.longitude,
          }
        : undefined,
      checkInTime: raw.check_in_time,
      checkOutTime: raw.check_out_time,
      amenities: raw.amenities || [],
      freeCancellation: raw.free_cancellation,
      images: (raw.images || []).map((img: any) => ({
        thumbnail: img.thumbnail,
        original: img.original_image,
      })),
      bookingLink: raw.serpapi_property_details_link,
      nearbyPlaces: (raw.nearby_places || []).map((np: any) => ({
        name: np.name,
        transportations: np.transportations,
      })),
    };
  }
}

File: backend-python/app/services/serpapi_provider.py
"""SerpApi provider — flights + hotels search via Google SERP."""

import httpx

from app.config import settings
from app.utils.logger import logger


class SerpApiProvider:
    BASE_URL = "https://serpapi.com/search"

    def _api_key(self) -> str:
        return settings.serpapi_api_key

    async def search_flights(
        self,
        origin: str,
        destination: str,
        departure_date: str,
        return_date: str | None = None,
        adults: int = 1,
        children: int = 0,
        travel_class: str = "economy",
        max_price: int | None = None,
        currency: str = "USD",
        deep_search: bool = False,
    ) -> list[dict]:
        if not self._api_key():
            logger.warning("SERPAPI_API_KEY not configured")
            return []

        params = {
            "engine": "google_flights",
            "departure_id": origin,
            "arrival_id": destination,
            "gl": "us",
            "hl": "en",
            "currency": currency,
            "type": "round_trip" if return_date else "one_way",
            "outbound_date": departure_date,
            "adults": adults,
            "children": children,
            "travel_class": travel_class,
            "api_key": self._api_key(),
        }
        if return_date:
            params["return_date"] = return_date
        if max_price:
            params["max_price"] = max_price

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                resp = await client.get(self.BASE_URL, params=params)
                data = resp.json()
        except Exception as e:
            logger.error(f"Flight search failed: {e}")
            return []

        flights = []
        is_round_trip = bool(return_date)
        for item in data.get("best_flights", []) + data.get("other_flights", []):
            all_legs = []
            for leg in item.get("flights", []):
                all_legs.append({
                    "departureAirport": {
                        "code": leg.get("departure_airport", {}).get("id", ""),
                        "name": leg.get("departure_airport", {}).get("name", ""),
                        "time": leg.get("departure_airport", {}).get("time", ""),
                    },
                    "arrivalAirport": {
                        "code": leg.get("arrival_airport", {}).get("id", ""),
                        "name": leg.get("arrival_airport", {}).get("name", ""),
                        "time": leg.get("arrival_airport", {}).get("time", ""),
                    },
                    "airline": leg.get("airline", ""),
                    "flightNumber": leg.get("flight_number", ""),
                    "duration": leg.get("duration", 0),
                    "airplane": leg.get("airplane", ""),
                    "travelClass": leg.get("travel_class", ""),
                    "overnight": leg.get("overnight", False),
                })

            # For round-trip, split legs into outbound and return groups.
            # The split point is the first leg that departs from the destination
            # airport (i.e., the return journey begins).
            outbound_legs = all_legs
            return_legs: list[dict] = []
            if is_round_trip and len(all_legs) > 1:
                dest_code = destination  # arrival_id passed by caller
                for idx, leg in enumerate(all_legs):
                    dep_code = leg.get("departureAirport", {}).get("code", "")
                    if dep_code and dep_code.upper() == dest_code.upper() and idx > 0:
                        outbound_legs = all_legs[:idx]
                        return_legs = all_legs[idx:]
                        break

            flight = {
                "id": item.get("flight_id", ""),
                "legs": all_legs,  # keep flat list for backwards compat
                "outboundLegs": outbound_legs,
                "returnLegs": return_legs,
                "layovers": [],
                "totalDuration": item.get("total_duration", 0),
                "price": item.get("price", 0),
                "currency": currency,
                "type": "round-trip" if return_date else "one-way",
                "isBest": item in data.get("best_flights", []),
                "bookingLink": item.get("booking_token", ""),
            }
            for leg in item.get("flights", []):
                if leg.get("layovers"):
                    for layover in leg["layovers"]:
                        flight["layovers"].append({
                            "airportCode": layover.get("airport", {}).get("id", ""),
                            "duration": layover.get("duration", 0),
                            "overnight": layover.get("overnight", False),
                        })

            flights.append(flight)

        return flights

    async def search_hotels(
        self,
        destination: str,
        check_in: str,
        check_out: str,
        adults: int = 2,
        children: int = 0,
        min_price: int | None = None,
        max_price: int | None = None,
        currency: str = "USD",
        sort: str = "relevance",
    ) -> list[dict]:
        if not self._api_key():
            logger.warning("SERPAPI_API_KEY not configured")
            return []

        params = {
            "engine": "google_hotels",
            "q": destination,
            "check_in_date": check_in,
            "check_out_date": check_out,
            "adults": adults,
            "children": children,
            "currency": currency,
            "gl": "us",
            "hl": "en",
            "sorting": sort,
            "api_key": self._api_key(),
        }
        if min_price:
            params["min_price"] = min_price
        if max_price:
            params["max_price"] = max_price

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                resp = await client.get(self.BASE_URL, params=params)
                data = resp.json()
        except Exception as e:
            logger.error(f"Hotel search failed: {e}")
            return []

        hotels = []
        for item in data.get("properties", []):
            hotels.append({
                "id": item.get("property_token", ""),
                "name": item.get("name", ""),
                "type": item.get("type", ""),
                "ratePerNight": item.get("rate_per_night", {}).get("extracted_lowest"),
                "totalRate": item.get("total_rate", {}).get("extracted_lowest"),
                "currency": currency,
                "rating": item.get("overall_rating"),
                "reviewsCount": item.get("reviews"),
                "amenities": item.get("amenities", []),
                "bookingLink": item.get("booking_token", ""),
                "images": [{"thumbnail": img.get("thumbnail"), "original": img.get("original")} for img in item.get("images", [])],
            })

        return hotels

    async def autocomplete(self, term: str) -> list[dict]:
        if not self._api_key():
            return []

        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                resp = await client.get(self.BASE_URL, params={
                    "engine": "google_autocomplete",
                    "q": term,
                    "api_key": self._api_key(),
                })
                data = resp.json()
        except Exception:
            return []

        suggestions = []
        for item in data.get("suggestions", []):
            suggestions.append({
                "code": item.get("value", ""),
                "name": item.get("value", ""),
            })
        return suggestions


serpapi_provider = SerpApiProvider()
